sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -0,0 +1,455 @@
+            dim0=nb_centroids_private_copies,
+            dim1=n_features,
+            dim2=n_clusters,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reduce_centroid_counts_private_copies = make_sum_reduction_2dim_kernel(
+            n_samples=n_clusters,
+            n_features=nb_centroids_private_copies,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reduce_centroids_private_copies = make_sum_reduction_3dim_kernel(
+            dim0=nb_centroids_private_copies,
+            dim1=n_features,
+            dim2=n_clusters,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        fixed_window_assignment_kernel = make_assignment_fixed_window_kernel(
+            n_samples,
+            n_features,
+            n_clusters,
+            preferred_work_group_size_multiple=self.preferred_work_group_size_multiple,
+            centroids_window_width_multiplier=self.centroids_window_width_multiplier,
+            centroids_window_height=self.centroids_window_height,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        # Transfer the input data to device memory,
+        # TODO: let the user pass directly dpctl.tensor or dpnp arrays to avoid copies.
+        if self.X_layout == "C":
+            # TODO: support the C layout and benchmark it and default to it if
+            # performances are better
+            raise ValueError("C layout is currently not supported.")
+            X_t = dpctl.tensor.from_numpy(X, device=self.device).T
+            assert (
+                X_t.strides[0] == 1
+            )  # Fortran memory layout, equivalent to C layout on transposed
+        elif self.X_layout == "F":
+            X_t = dpctl.tensor.from_numpy(X.T, device=self.device)
+            assert (
+                X_t.strides[1] == 1
+            )  # C memory layout, equivalent to Fortran layout on transposed
+        else:
+            raise ValueError(
+                f"Expected X_layout to be equal to 'C' or 'F', but got {self.X_layout} ."
+            )
+
+        centroids_t = dpctl.tensor.from_numpy(centers_init.T, device=self.device)
+
+        # Allocate the necessary memory in the device global memory
+        centroids_t_copy_array = dpctl.tensor.empty_like(
+            centroids_t, device=self.device
+        )
+        best_centroids_t = dpctl.tensor.empty_like(
+            centroids_t_copy_array, device=self.device
+        )
+        centroids_half_l2_norm = dpctl.tensor.empty(
+            n_clusters, dtype=dtype, device=self.device
+        )
+        centroid_counts = dpctl.tensor.empty(
+            n_clusters, dtype=dtype, device=self.device
+        )
+        center_shifts = dpctl.tensor.empty(n_clusters, dtype=dtype, device=self.device)
+        inertia = dpctl.tensor.empty(n_samples, dtype=dtype, device=self.device)
```

Comment:
This should be renamed to `per_sample_inertia` to make the code of the kernel easier to follow.